inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/HookManager.java

```diff
@@ -0,0 +1,221 @@
+        Duration delay = conf.getInterBatchDelay();
+
+        classInstrumentationJob = executor.startJob(this::checkClassesForConfigurationUpdates, conf.getClassConfigurationCheckBatchSize(), delay, delay);
+    }
+
+    @PreDestroy
+    void destroy() {
+        Instances.hookManager = NoopHookManager.INSTANCE;
+        classInstrumentationJob.cancel();
+    }
+
+
+    private IMethodHook getHook(Class<?> clazz, String methodSignature) {
+        try {
+            val hook = hooks.get(clazz).get(methodSignature);
+            return hook == null ? NoopMethodHook.INSTANCE : hook;
+        } catch (ExecutionException e) {
+            //Should in theory never occur, as we only call new ConcurrentHashMap() in the cache loader
+            log.error("Error in cache", e);
+            return NoopMethodHook.INSTANCE;
+        }
+    }
+
+
+    @Override
+    public void onNewClassesDiscovered(Set<Class<?>> newClasses) {
+        for (Class<?> clazz : newClasses) {
+            pendingClasses.put(clazz, Boolean.TRUE);
+        }
+        selfMonitorQueueSize();
+    }
+
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+        InternalSettings newInternal = ev.getNewConfig().getInstrumentation().getInternal();
+        classInstrumentationJob.setBatchSizes(newInternal.getClassConfigurationCheckBatchSize());
+        classInstrumentationJob.setInterBatchDelay(newInternal.getInterBatchDelay());
+    }
+
+    @EventListener
+    private void instrumentationConfigEventListener(InstrumentationConfigurationChangedEvent ev) {
+        for (Class<?> clazz : instrumentation.getAllLoadedClasses()) {
+            pendingClasses.put(clazz, Boolean.TRUE);
+        }
+        selfMonitorQueueSize();
+    }
+
+    private void checkClassesForConfigurationUpdates(int batchSize) {
+
+        Stopwatch watch = Stopwatch.createStarted();
+
+        int checkedClassesCount = 0;
+
+        Iterator<Class<?>> queueIterator = pendingClasses.asMap().keySet().iterator();
+        while (queueIterator.hasNext() && checkedClassesCount < batchSize) {
+            Class<?> clazz = queueIterator.next();
+            queueIterator.remove();
+            try {
+                updateHooksForClass(clazz);
+            } catch (Throwable e) {
+                log.error("Error checking for hook configuration updates for {}", clazz.getName(), e);
+            }
+            checkedClassesCount++;
+        }
+        if (checkedClassesCount > 0) {
+            log.debug("Checked hook configuration of {} classes in {} ms, {} classes left to check",
+                    checkedClassesCount, watch.elapsed(TimeUnit.MILLISECONDS), pendingClasses.size());
+        }
+
+        selfMonitorQueueSize();
+    }
+
+    private void updateHooksForClass(Class<?> clazz) {
+        Map<MethodDescription, MethodHookConfiguration> hookConfigs = configResolver.getHookConfigurations(clazz);
+
+        ConcurrentHashMap<String, MethodHook> activeHooks;
```

Comment:
rename thos `activeClassHooks`, this took me 15 mins to understand that active hooks are not all hooks, but only the one related to this class...